primitives/src/crhf.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Collision-resistant Hash Functions (CRHF) definitions and implementations.
+
+use ark_std::{
+    borrow::Borrow,
+    fmt::Debug,
+    hash::Hash,
+    marker::PhantomData,
+    rand::{CryptoRng, RngCore},
+    string::ToString,
+    vec::Vec,
+};
+
+use crate::{
+    errors::PrimitivesError,
+    rescue::{sponge::RescueCRHF, RescueParameter, CRHF_RATE},
+};
+
+/// A trait for CRHF
+/// (based on ark-primitives' definition, but self-declared for minimal
+/// dependency and easier future upgradability.)
+pub trait CRHF {
+    /// Input to the CRHF
+    type Input;
+    /// Output of the CRHF
+    // FIXME: (alex) wait until arkwork 0.4.0 to add the following:
+    // + Default + CanonicalSerialize + CanonicalDeserialize;
+    // right now, const-generic are not supported yet.
+    type Output: Clone + PartialEq + Eq + Hash + Debug;
+
+    /// Public parameters of the CRHF (if any)
+    type Parameters: Clone + Debug + Sync + Send;
+
+    /// picking the concrete CRHF from the parameterized hash family
+    fn setup<R: RngCore + CryptoRng>(rng: &mut R) -> Result<Self::Parameters, PrimitivesError>;
```

Comment:
I wouldn't define it if it's not used at all in the current impl.
If we were to take arkworks traits and implement these, then sure, we need to have the setup, but since (for now at least) we're defining our own trait, I think it's ok to leave it out, especially that (potentially) adding it back once we move to arkwork's traits is going to be trivial

Reply:
good point! will remove for now.